### PR TITLE
Copy passage with verse numbers #171

### LINKED ISSUE #171: Share bible passage should include verse numbers
From user:

> When you copy and paste a passage, it no longer displays the verse number. It would be really great if you could add that.

* perhaps this should be optional (settings)

Implementation details:
 - Inherit class from OsisToCanonicalTextHandler, implement verse/chapter numbers there and use it instead when sharing/copying. 
 - Remember localisation (similarly as in speaking), i.e. string "chapter" should be in the same language as the bible text (not in interface language)

### PR DESCRIPTION
**Pull request by timbze**
Closes #171
 Include verse numbers in copy or share Bible passage.
Disable feature by disabling "Show Verse Numbers" feature for Bible text.

### CODE REVIEW

File: app/src/main/java/net/bible/service/sword/SwordContentFacade.java [hunk 1]
```diff
@@ -231,8 +231,8 @@ public String getCanonicalText(Book book, Key key) throws NoSuchKeyException, Bo
     	try {
 			BookData data = new BookData(book, key);
 			SAXEventProvider osissep = data.getSAXEventProvider();
-		
-			ContentHandler osisHandler = new OsisToCanonicalTextSaxHandler();
+
+			ContentHandler osisHandler = new OsisToCopyTextSaxHandler(key.getOsisID().trim().contains(" "));
```
Comment (tuomas2) [inline app/src/main/java/net/bible/service/sword/SwordContentFacade.java]: Hmm, can a single key ever contain multiple verses and space?  

Could this Key be also VerseRange? Then there could be also verse ranges like Gen1.1-Gen.1.2

File: app/src/main/java/net/bible/service/sword/SwordContentFacade.java [hunk 2]
```diff
@@ -231,8 +231,8 @@ public String getCanonicalText(Book book, Key key) throws NoSuchKeyException, Bo
     	try {
 			BookData data = new BookData(book, key);
 			SAXEventProvider osissep = data.getSAXEventProvider();
-		
-			ContentHandler osisHandler = new OsisToCanonicalTextSaxHandler();
```
[the remark this answers is not in the document]
Comment (tuomas2) [reply app/src/main/java/net/bible/service/sword/SwordContentFacade.java]: So please trace back (with debugger etc.) from where this function is called when share/copy function is triggered and use another new function for that.

[the remark this answers is not in the document]
Comment (timbze) [reply app/src/main/java/net/bible/service/sword/SwordContentFacade.java]: I did a usage search on the function but didn't look deep enough. You're right it should be a new function. Done now

File: app/src/main/java/net/bible/service/format/osistohtml/osishandlers/OsisToCopyTextSaxHandler.kt [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+    private var verseCount: Int = 0
+    private var writeChapter: String = ""
+    private val showVerseNumbersSetting: Boolean = CommonUtils.getSharedPreferences().getBoolean("show_verseno_pref", true)
+
+    /*
+     * Called when the starting of the Element is reached. For Example if we have Tag
+     * called <Title> ... </Title>, then this method is called when <Title> tag is
+     * Encountered while parsing the Current XML File. The AttributeList Parameter has
+     * the list of all Attributes declared for the Current Element in the XML File.
+    */
```
[the remark this answers is not in the document]
Comment (timbze) [reply app/src/main/java/net/bible/service/format/osistohtml/osishandlers/OsisToCopyTextSaxHandler.kt]: Done

File: app/src/main/java/net/bible/service/format/osistohtml/osishandlers/OsisToCopyTextSaxHandler.kt [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+                if (attrs != null && showVerseNumbersSetting) {
+                    val oldChapterNumber = currentChapterNumber
+                    currentChapterNumber = TagHandlerHelper.osisIdToVerseNum(attrs.getValue("", OSISUtil.OSIS_ATTR_OSISID))
+                    if (oldChapterNumber > 0 && currentChapterNumber > oldChapterNumber) {
+                        writeChapter = "$currentChapterNumber:"
+                    }
+                }
+            }
+            OSISUtil.OSIS_ELEMENT_VERSE -> {
+                verseCount++
```
[the remark this answers is not in the document]
Comment (timbze) [reply app/src/main/java/net/bible/service/format/osistohtml/osishandlers/OsisToCopyTextSaxHandler.kt]: Good catch :) I did it at first to not add verse number on first verse, but decided to put verse number at each verse, then forgot to remove. Done

File: app/src/main/java/net/bible/service/sword/SwordContentFacade.java (hunk 1 above)
Comment (timbze) [reply app/src/main/java/net/bible/service/sword/SwordContentFacade.java]: From my testing when multiple verses are selected, the key value becomes "Gen.1.1 Gen.1.2 Gen.1.3" etc. I couldn't find another easy way yet to get verse count from key but maybe you know how. I actually find that the calling function shareVerse (PageControl.java line 97) passes a verseRange object to the key.
I could use verseRange and use verseRange.toVerseArray().length, but don't know what is the best way.

File: app/src/test/java/net/bible/service/format/osistohtml/osishandlers/OsisToCopyTextSaxHandlerTest.kt [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+import org.junit.Test
+
+import org.hamcrest.CoreMatchers.equalTo
+import org.junit.Assert.assertThat
+import org.xml.sax.helpers.AttributesImpl
+
+class OsisToCopyTextSaxHandlerTest {
+
+    private var osisToCopyTextSaxHandlerSingleVerse: OsisToCopyTextSaxHandler? = null
+    private var osisToCopyTextSaxHandlerMultipleVerses: OsisToCopyTextSaxHandler? = null
```
Comment (tuomas2) [inline app/src/test/java/net/bible/service/format/osistohtml/osishandlers/OsisToCopyTextSaxHandlerTest.kt]: I think you could use lateinit variables

Comment (timbze) [reply app/src/test/java/net/bible/service/format/osistohtml/osishandlers/OsisToCopyTextSaxHandlerTest.kt]: Done.

**Pull request merged**